src/coreclr/src/jit/morph.cpp

```diff
@@ -18451,38 +18471,91 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
-                assert(fieldSeq->GetTail()->m_fieldHnd == indir->AsField()->gtFldHnd);
+                // Skip TYP_STRUCT IND nodes, it's not clear what we can do with them.
+                // Normally these should appear only as sources of variable sized copy block
+                // operations (DYN_BLK) so it probably doesn't make much sense to try to
+                // convert these to local nodes.
+                return;
             }
+
+            if ((user == nullptr) || !user->OperIs(GT_ASG))
+            {
+                // TODO-ADDR: Skip TYP_STRUCT indirs for now, unless they're used by an ASG.
+                // At least call args will require extra work because currently they must be
+                // wrapped in OBJ nodes so we can't replace those with local nodes.
+                return;
+            }
+
+            if (indir->OperIs(GT_FIELD))
+            {
+                CORINFO_CLASS_HANDLE fieldClassHandle;
+                CorInfoType          corType =
+                    m_compiler->info.compCompHnd->getFieldType(indir->AsField()->gtFldHnd, &fieldClassHandle);
+                assert(corType == CORINFO_TYPE_VALUECLASS);
+
+                structLayout = m_compiler->typGetObjLayout(fieldClassHandle);
+            }
+            else
+            {
+                structLayout = indir->AsBlk()->GetLayout();
+            }
+
+            // We're not going to produce a TYP_STRUCT LCL_FLD so we don't need the field sequence.
+            fieldSeq = nullptr;
         }
 
-        indir->ChangeOper(GT_LCL_FLD);
-        indir->AsLclFld()->SetLclNum(val.LclNum());
-        indir->AsLclFld()->SetLclOffs(val.Offset());
-        indir->AsLclFld()->SetFieldSeq(fieldSeq == nullptr ? FieldSeqStore::NotAField() : fieldSeq);
+        // We're only processing TYP_STRUCT variables now so the layout should never be null,
+        // otherwise the below layout equality check would be insufficient.
+        assert(varDsc->GetLayout() != nullptr);
+
+        if ((val.Offset() == 0) && (structLayout == varDsc->GetLayout()))
+        {
+            indir->ChangeOper(GT_LCL_VAR);
+            indir->AsLclVar()->SetLclNum(val.LclNum());
+        }
+        else if (!varTypeIsStruct(indir->TypeGet()))
+        {
+            indir->ChangeOper(GT_LCL_FLD);
+            indir->AsLclFld()->SetLclNum(val.LclNum());
+            indir->AsLclFld()->SetLclOffs(val.Offset());
+            indir->AsLclFld()->SetFieldSeq(fieldSeq == nullptr ? FieldSeqStore::NotAField() : fieldSeq);
+
+            // Promoted struct vars aren't currently handled here so the created LCL_FLD can't be
+            // later transformed into a LCL_VAR and the variable cannot be enregistered.
+            m_compiler->lvaSetVarDoNotEnregister(val.LclNum() DEBUGARG(Compiler::DNER_LocalField));
+        }
+        else
+        {
+            // TODO-ADDR: Add TYP_STRUCT support to LCL_FLD.
+            return;
```

Comment:
@mikedn How far are we from having TYP_STRUCT LCL_FIELDs? Is this something you are working on? If not, do you know what the main blockers for enabling this are?

Reply:
@erozenfeld I'm working on it in my own runtime fork. I do not plan to work on new dotnet PRs, except to fix bugs in code I have wrote in the past.

One problem is that the codegen for struct `PUTARG_STK` currently only supports `LCL_VAR` and `OBJ` source nodes. The code is messy and not that easy to change. However, it should be possible to avoid changing this code in the beginning - just wrap struct `LCL_FLD`s in `OBJ`s in lowering.

Another more serious problem is `fgMorphCopyBlock`, this one has to be updated and it's a large and messy piece of code.

VN may need some small updates but shouldn't be a problem. Same for liveness, though that code is pretty clunky as it is and may benefit from a more serious shake up.

And of course, once you have struct `LCL_FLD`s there's no much need to keep `IsLocalAddrExpr` & friends. That's a lot of stuff to cleanup.

Reply:
`I do not plan to work on new dotnet PRs, except to fix bugs in code I have wrote in the past.`

:(:(:(

Reply:
@mikedn Thank you for a detailed reply. I'm looking at reducing the number of cases where we mark locals as address-exposed unnecessarily and this was one of the places we'll need to fix.

Reply:
@erozenfeld I don't follow. Struct typed field access does not make variables address exposed even without struct typed `LCL_FLD`s. Yes, you get trees like `OBJ(ADDR(LCL_FLD))` but the variable isn't marked address exposed.

I'm not sure there's a lot that can be done to reduce the number of address exposed vars:
* Variables passed by ref in calls are address exposed, not a lot that can be done about that without interprocedural analysis. And this probably the most common case...
* Sometimes user code does exotic things like reinterpretation and in this case `LocalAddressVisitor` & morph are unnecessarily conservative. Many cases can be avoided by simply using `LCL_FLD` or `BITCAST` to perform reinterpretation. I have an experiment that uses `BITCAST` to avoid cases like `*(int*)&floatVar` and it caught a few cases in corelib but I don't think this is common.
* Other exotic case that `LocalAddressVisitor` is missing is pointer arithmetic (`ADD`, `SUB`, maybe relops). There's actually an open issue about handling something like `&lcl.f - &lcl` as a fast way to obtain the offset of a field but again, this isn't common.
* There's a pesky pattern that the C# compiler generates, I don't remember the exact details but sometimes it transforms code like `x.y.z++` into:
```
int* a = &x.y.z;
int t = *a;
t = t + 1;
*a = t;
```
thus forcing the address to be used across trees, something that `LocalAddressVisitor` doesn't currently handle. This is something I'd like to look at one day, it should be possible to track `LocalAddressVisitor` "values" across trees. And the recent movement of predecessor computation earlier in phase order might help tracking "values" even across blocks.
And of course, thanks to the introduction of `ref` variables in C#, users themselves can write similar code.
* There's also some cases where the JIT shoots itself in the foot by making variables address exposed for no good reason. One case that I know about is `fgMorphCopyBlock` field by field copying. If memory serves me well, just fixing that produces a 20kbytes improvement in fx diff.